### PR TITLE
[BEAM-1195, BEAM-115] Add full-fidelity conversion to/from Java SDK triggers and Runner API protos

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This PR has somewhat independent commits that make sense to review together:

1. Alter the Java SDK's trigger classes to have the necessary functionality to reconstruct easily from a serialized form.
2. Simplify some inheritances relationship in the Java SDK that were actually incorrect.
3. Flesh out the Runner API proto to include all the needed bits to serialize triggers fully.
4. Actually serialize triggers and their inverse; test the round trip.

### CODE REVIEW

**Comment by kennknowles** [comment]
This will require a new worker image from Dataflow. So please review even in the presence of Jenkins failure.

**Comment by kennknowles** [comment]
R: @tgroh

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  abstract static class ToProtoAndBackSpec {
+    abstract Trigger getTrigger();
+  }
+
+  private static ToProtoAndBackSpec toProtoAndBack(Trigger trigger) {
+    return new AutoValue_TriggersTest_ToProtoAndBackSpec(trigger);
+  }
+
+  @Parameters(name = "{index}: {0}")
+  public static Iterable<Object[]> data() {
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java]: You can just provide an `Iterable<Object>`, since you only take a single argument.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+/** Tests for utilities in {@link Triggers}. */
+@RunWith(Parameterized.class)
+public class TriggersTest {
+
+  @AutoValue
+  abstract static class ToProtoAndBackSpec {
+    abstract Trigger getTrigger();
+  }
+
+  private static ToProtoAndBackSpec toProtoAndBack(Trigger trigger) {
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java]: This is written to sound like the test method. Add "`Spec`" to the end?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterPaneStateMachine.java [hunk 3]
```diff
@@ -40,24 +40,24 @@
       StateTags.makeSystemTagInternal(StateTags.combiningValueFromInputInternal(
           "count", VarLongCoder.of(), Sum.ofLongs()));
 
-  private final int countElems;
+  private final long countElems;
 
-  private AfterPaneStateMachine(int countElems) {
+  private AfterPaneStateMachine(long countElems) {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterPaneStateMachine.java]: Changed back to int, and shrunk the value in the proto to make it clear. Note that we could check for overflow, but we actually want to do a blind write to the accumulator.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/TriggerStateMachines.java [hunk 4]
```diff
@@ -171,18 +174,38 @@ private TriggerStateMachine evaluateSpecific(OrFinallyTrigger v) {
...
     private static class AfterDelayFromFirstElementStateMachineAdapter
         extends AfterDelayFromFirstElementStateMachine {
 
-      public AfterDelayFromFirstElementStateMachineAdapter(AfterDelayFromFirstElement v) {
-        this(v.getTimeDomain(), v.getTimestampMappers());
+      private static final Function<TimestampTransform, SerializableFunction<Instant, Instant>>
+          CONVERT_TIMESTAMP_TRANSFORM =
+              new Function<TimestampTransform, SerializableFunction<Instant, Instant>>() {
+                @Override
+                @Nonnull
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/TriggerStateMachines.java]: Isn't it for use when overriding a `@Nullable` thing? In the case of the parameter, it indicates that it must not be null, despite the signature (actually an unsafe alteration) while in the case of the return value it indicates an additional guarantee that it won't return null.

TBH this way of hacking nullability onto the type system without real integration just doesn't quite work. I'll happily remove this one.

File: sdks/common/runner-api/src/main/proto/beam_runner_api.proto [hunk 5]
```diff
@@ -402,6 +402,21 @@ enum OutputTime {
...
+
+  // Event time is time from the perspective of the data
+  EVENT_TIME = 0;
+
+  // Processing time is time from the perspective of the
+  // execution of your pipeline
+  PROCESSING_TIME = 1;
+
+  // Synchronized processing time is the minimum of the
+  // processing time of all upstream dependencies
```
Comment (kennknowles) [inline sdks/common/runner-api/src/main/proto/beam_runner_api.proto]: Changed to your wording + explanatory follow-up

File: sdks/common/runner-api/src/main/proto/beam_runner_api.proto [hunk 6]
```diff
@@ -475,20 +491,53 @@ message Trigger {
...
 
   // The full disjoint union of possible triggers.
   oneof trigger {
     AfterAll after_all = 1;
     AfterAny after_any = 2;
     AfterEach after_each = 3;
     AfterEndOfWindow after_end_of_widow = 4;
-    AfterProcessingTimeDelay after_processing_time_delay = 5;
-    AfterSynchronizedProcessingTimeDelay after_synchronized_processing_time_delay = 6;
+    AfterProcessingTime after_processing_time = 5;
```
Comment (kennknowles) [inline sdks/common/runner-api/src/main/proto/beam_runner_api.proto]: Alphabetized. Since nothing uses this API, also put the indices in order. I'll stop that in a couple weeks.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java [hunk 7]
```diff
@@ -17,40 +17,87 @@
...
- *
- * <p>The time at which to fire the timer can be adjusted via the methods in
- * {@link AfterDelayFromFirstElement}, such as {@link AfterDelayFromFirstElement#plusDelayOf} or
- * {@link AfterDelayFromFirstElement#alignedTo}.
  */
 @Experimental(Experimental.Kind.TRIGGER)
-public class AfterProcessingTime extends AfterDelayFromFirstElement {
+public class AfterProcessingTime extends OnceTrigger {
+
+  private static final PeriodFormatter PERIOD_FORMATTER = PeriodFormat.wordBased(Locale.ENGLISH);
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java [hunk 8]
```diff
@@ -17,40 +17,87 @@
...
+  public AfterProcessingTime plusDelayOf(final Duration delay) {
+    return new AfterProcessingTime(
+        ImmutableList.<TimestampTransform>builder()
+            .addAll(timestampTransforms)
+            .add(TimestampTransform.delay(delay))
+            .build());
+  }
+
+  /**
+   * Aligns timestamps to the smallest multiple of {@code size} since the {@code offset} greater
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java]: Yes, `new Instant(0)` is Java for "the timestamp since the epoch" but I will clarify since my wording was really bad.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterSynchronizedProcessingTime.java [hunk 9]
```diff
@@ -18,21 +18,22 @@
...
+ * first element's arrival.
+ *
+ * <p>This is for internal, primarily as a "continuation trigger" for {@link AfterProcessingTime}
+ * triggers. In that use, this trigger is ready as soon as all upstream workers processing time
+ * clocks have caught up to the moment that input arrived.
  */
-public class AfterSynchronizedProcessingTime extends AfterDelayFromFirstElement {
+public class AfterSynchronizedProcessingTime extends OnceTrigger {
 
   public AfterSynchronizedProcessingTime() {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterSynchronizedProcessingTime.java]: Done. Been that way for a while, but hey as usual a static factory method clarifies things.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+
+  @VisibleForTesting
+  static class ProtoConverter {
+
+    public RunnerApi.Trigger convertTrigger(Trigger trigger) {
+      Method evaluationMethod = getEvaluationMethod(trigger.getClass());
+      return tryConvert(evaluationMethod, trigger);
+    }
+
+    public RunnerApi.Trigger evaluateOnceTrigger(OnceTrigger trigger) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: Deleted. We don't actually need it here. For state machines, at the moment, we keep track of once vs non-once triggers. But I don't think tracking that statically is all that useful, and did not include the distinction in the proto. The converse is much more important (does this trigger finish?) and we don't track that. So just treating all of that as a dynamic property is where I am going.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+        throw new IllegalStateException(
+            String.format("Internal error: could not invoke %s", evaluationMethod));
+      }
+    }
+
+    private Method getEvaluationMethod(Class<?> clazz) {
+      try {
+        return getClass().getDeclaredMethod("convertSpecific", clazz);
+      } catch (NoSuchMethodException exc) {
+        throw new UnsupportedOperationException(
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+                      new Instant(transform.getAlignTo().getOffset()));
+              break;
+            case DELAY:
+              trigger = trigger.plusDelayOf(Duration.millis(transform.getDelay().getDelayMillis()));
+              break;
+            case TIMESTAMPTRANSFORM_NOT_SET:
+            default:
+              throw new IllegalArgumentException(
+                  String.format(
+                      "Unknown timestamp transform case: %s",
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java (hunk 2 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java (hunk 1 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/TriggersTest.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java [hunk 13]
```diff
@@ -60,17 +107,28 @@ public Instant getWatermarkThatGuaranteesFiring(BoundedWindow window) {
...
-    for (SerializableFunction<Instant, Instant> delayFn : timestampMappers) {
-      builder
-          .append(".plusDelayOf(")
-          .append(delayFn)
-          .append(")");
+    for (TimestampTransform transform : getTimestampTransforms()) {
+      if (transform instanceof TimestampTransform.Delay) {
+        TimestampTransform.Delay delay = (TimestampTransform.Delay) transform;
+        builder
+            .append(".plusDelayOf(")
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java]: It feels that this should just be a defined `TimestampTransform.toString()`, and this method gets a whole lot simpler.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterSynchronizedProcessingTime.java [hunk 14]
```diff
@@ -18,21 +18,26 @@
...
 
 /**
- * A trigger that fires after synchronized processing time has reached a shared
- * threshold between upstream workers.
+ * A trigger that fires after synchronized processing time has reached the processing time of the
+ * first element's arrival.
+ *
+ * <p>This is for internal, primarily as a "continuation trigger" for {@link AfterProcessingTime}
+ * triggers. In that use, this trigger is ready as soon as all upstream workers processing time
+ * clocks have caught up to the moment that input arrived.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterSynchronizedProcessingTime.java]: Add something to strongly discourage use by users?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.windowing;
+
+import com.google.auto.value.AutoValue;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** An abstract description of a standardized transformation on timestamps. */
+public abstract class TimestampTransform {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java]: Can these actually be inlined into `AfterProcessingTime`? I believe it is very likely.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** Utilities for working with {@link Triggers Triggers}. */
+@Experimental(Experimental.Kind.TRIGGER)
+public class Triggers implements Serializable {
+
+  @VisibleForTesting static final ProtoConverter CONVERTER = new ProtoConverter();
+
+  public static RunnerApi.Trigger triggerToProto(Trigger trigger) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: just `toProto` seems as though it should be sufficient. Alternatively, `convertToProto`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+  static class ProtoConverter {
+
+    public RunnerApi.Trigger convertTrigger(Trigger trigger) {
+      Method evaluationMethod = getEvaluationMethod(trigger.getClass());
+      return tryConvert(evaluationMethod, trigger);
+    }
+
+    private RunnerApi.Trigger tryConvert(Method evaluationMethod, Trigger trigger) {
+      try {
+        return (RunnerApi.Trigger) evaluationMethod.invoke(this, trigger);
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: I kind of like the style of having a map of `<Class<? extends Trigger>, TriggerConverter>` where the latter is `TriggerConverter<T extends Trigger>`, with `convert(T)`, the same way we do override factories. The actual invocations are not really typesafe, but you don't have to write all the casting code yourself.

Thoughts?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterSynchronizedProcessingTimeStateMachine.java [hunk 18]
```diff
@@ -27,13 +27,17 @@
 
 class AfterSynchronizedProcessingTimeStateMachine extends AfterDelayFromFirstElementStateMachine {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterSynchronizedProcessingTimeStateMachine.java]: Yea, I was just trying to keep this PR manageably sized. Since the state machines are basically independent of the proto concerns, I'd prefer not to mess with them. There will certainly be a follow-up PR where a state machine is built from a proto instead of an SDK trigger. It is not mandatory, since we could parse the proto to an SDK trigger then go from there to a state machine, but it seems more faithful to the model. And it will be net negative LOC. All that is to say that when I say "save it for later" it is a real time that will come, not an excuse to not do it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java (hunk 13 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java]: Agreed. The fact that triggers try to `toString()` like their static factory methods is not something I care deeply about. But the logic I've added here really is just to keep parity (modulo the fact that the code wasn't really complete previously). OK to defer that? There will surely be another PR converting triggers to autovalue, which seems like a good time to do that kind of cleanup.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterSynchronizedProcessingTime.java (hunk 14 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterSynchronizedProcessingTime.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java (hunk 15 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java]: Do you making this an inner class? I don't think that adds value. They aren't defined with respect to `AfterProcessingTime`, and they don't depend on it for interpretation.

The issue of whether these will be used anywhere else is also not entirely settled.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java (hunk 16 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java (hunk 17 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java]: The casting code you have to write will be largely the same. When you pull a converter out of the map, it is the same as grabbing a method, and invoking requires the same sort of type-unsafe call.

The difference is really in writing a converter/visitor. It is very readable to write a bunch of overloads to cover the cases (somewhat like `DoFnSignature.Parameter.Cases`) but since anonymous functions are so bulky in Java 7 a map-based implementation adds a ton of readability overhead.

BTW the pattern here is loosely modeled on http://www.anzaan.com/2012/04/java-visitor-pattern-and-the-expression-problem-part-1-framework-based-approach/ and is also a modified version of the visitor logic from `TriggerStateMachines` (not shared because of dependency concerns). The reason for this approach is all about extensibility.

As it turns out, we currently don't have that many trigger syntax tree traversals, so I de-genericized the approach so it isn't really a reusable "TriggerEvaluator". That seemed like a bulky thing to put in the user-facing API. And perhaps we never will have more traversals, as we'll write them against the proto directly to make sure we cover all cases in the model, not just the Java SDK. I'm not sure.

My favorite solution as far as raw Java would be to make `Trigger` more like `DoFnSignature.Parameter` and then have its own `Cases`. Basically the same as a generic `TriggerEvaluator` but the terminology is a bit more obvious IMO. The backwards-compatibility breakage might start to be too much. We'd have to reconcile our seemingly open-ended fluent syntax for triggers with acknowledging that they are closed-world and have a centralized list of cases. I'm sure it is possible, just not taking it on right here.

**Comment by kennknowles** [comment]
@tgroh you've brought up a lot of good style and software engineering questions here; I didn't address nearly all of them. I want to emphasize that I agree that they are good ideas and I am not trying to avoid them, even though I actually _am_ trying to avoid lumping them into this PR.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterSynchronizedProcessingTimeStateMachine.java (hunk 18 above)
Comment (tgroh) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterSynchronizedProcessingTimeStateMachine.java]: Add a JIRA and a TODO linking to it and I'm happy.

**Comment by tgroh** [comment]
(on green, of course)

**Comment by kennknowles** [comment]
Run Dataflow RunableOnService
